lib/msprime.c

```diff
@@ -2116,6 +2119,37 @@ msp_common_ancestor_event(msp_t *self, population_id_t population_id)
 }
 
 
+/* This calculates the rate given by Eq (2) in the notes
+ */
+static double
+msp_compute_lambda_Xi_dirac(msp_t *self, unsigned int b)
+{
+    unsigned int l, n_block;
+    double psi = self->model.params.dirac_coalescent.psi;
+    double c = self->model.params.dirac_coalescent.const_c;
+    double ret = 0;
+
+    assert(b > 0);
+    assert(psi > 0);
+    assert(psi < 1);
+    assert(c >= 0);
+
+    n_block = GSL_MIN(b, 4);
+    for (l = 0; l < n_block; l++){
+        ret += gsl_sf_exp(gsl_sf_lnchoose(b, l) +
+                           gsl_sf_log(4.0) +
+                           (double)l * gsl_sf_log(psi/4.0) +
+                           (double)(b-l) * gsl_sf_log(1-psi));
+    }
+
+
+    ret = 1 - ret;
+    ret *= c*4.0/pow(psi,2.0);
+    ret += b * (b-1) / 2;
```

Comment:
We need to be careful to cast the intermediate values to a double here, as b * (b - 1) will overflow. It might be safer to do something like
```c
msp_compute_lambda_Xi_dirac(msp_t *self, unsigned int num_ancestors)
{
     double b = num_ancestors;
```

You can safely remove the two casts in the loop also, since (b - l) will be promoted to double.